e2e-tests/src/test/electing_validators.rs

```diff
@@ -0,0 +1,266 @@
+fn stakers_as_storage_keys<C: AnyConnection>(
+    connection: &C,
+    accounts: &[AccountId],
+    era: EraIndex,
+) -> BTreeSet<StorageKey> {
+    accounts
+        .iter()
+        .map(|acc| {
+            connection
+                .as_connection()
+                .metadata
+                .storage_double_map_key("Staking", "ErasStakers", era, acc)
+                .unwrap()
+        })
+        .collect()
+}
+
+/// Verify that `pallet_staking::ErasStakers` contain all target validators.
+///
+/// We have to do it by comparing keys in storage trie.
+fn assert_validators_are_elected_stakers<C: AnyConnection>(
+    connection: &C,
+    current_era: EraIndex,
+    expected_validators_as_keys: &BTreeSet<StorageKey>,
+) {
+    let storage_key = get_eras_stakers_storage_key(current_era);
+    let stakers = connection
+        .as_connection()
+        .get_keys(storage_key, None)
+        .unwrap_or_else(|_| panic!("Couldn't read storage keys"))
+        .unwrap_or_else(|| panic!("Couldn't read `ErasStakers` for era {}", current_era))
+        .into_iter()
+        .map(|key| {
+            let mut bytes = [0u8; 84];
+            hex::decode_to_slice(&key[2..], &mut bytes as &mut [u8]).expect("Should decode key");
+            StorageKey(bytes.to_vec())
+        });
+    let stakers = BTreeSet::from_iter(stakers);
+
+    assert_eq!(
+        *expected_validators_as_keys, stakers,
+        "Expected another set of staking validators.\n\tExpected: {:?}\n\tActual: {:?}",
+        expected_validators_as_keys, stakers
+    );
+}
+
+/// Verify that all target validators are included `pallet_session::Validators` across a few
+/// consecutive sessions.
+fn assert_validators_are_used_as_authorities<C: AnyConnection>(
+    connection: &C,
+    expected_authorities: &BTreeSet<AccountId>,
+) {
+    let mut authorities = BTreeSet::new();
+    // There are 4 slots, 3 reserved validators and 3 nonreserved, therefore after 3 sessions
+    // we should observe all of them.
+    for _ in 0..3 {
+        let current_session = get_current_session(connection);
+
+        info!("Reading authorities in session {}", current_session);
+        let current_authorities: Vec<AccountId> =
+            connection.read_storage_value("Session", "Validators");
+        for ca in current_authorities.into_iter() {
+            authorities.insert(ca);
+        }
+
+        wait_for_session(connection, current_session + 1).expect("Couldn't wait for next session");
+    }
+
+    assert_eq!(
+        *expected_authorities, authorities,
+        "Expected another set of authorities.\n\tExpected: {:?}\n\tActual: {:?}",
+        expected_authorities, authorities
+    );
+}
+
+/// Chill validator.
+fn chill(connection: &SignedConnection) {
+    let xt = compose_extrinsic!(connection.as_connection(), "Staking", "chill");
+    send_xt(
+        connection,
+        xt,
+        Some("chilling validator"),
+        XtStatus::InBlock,
+    );
+}
+
+/// Chill all validators in `malicious`.
```

Comment:
I would not use the word `malicious` as it is kinda misleading. They are doing what is allowed, nothing is evil about them :P

Reply:
yeah, right; renamed